[FEAT] Allow user to directly use a JWT token instead of logging in with username / password

anneclairelh opened the pull request:
This feature is needed for the Jupyter Notebook of PQD which will have the JWT token stored as an environment variable. 
Right now, it is kept as a alternative mechanism to the regular login.
In order to address the issue of token expiration, a temporary error is raised which suggests the user to perform an interactive login. Therefore, a new optional parameter is added to the SDK client, namely the `login_url` to be used.
@penguinoneshaw  let me know what you think about this implementation.

awennersteen commented:
Changelog needs update

MatthieuMoreau0 commented on sdk/errors.py:
```diff
@@ -39,3 +39,7 @@ def __str__(self) -> str:
         )
+
+
+class CannotLoginError(Exception):
```
`LoginError` seems sufficient for the naming

MatthieuMoreau0 commented on sdk/client.py:
```diff
@@ -82,6 +85,11 @@ def _request(
         # If account returns unauthorized we attempt to login and retry request
         if rsp.status_code == 401:
+            #  TODO: Improve the interactive login flow
+            if (self.username == "" and self.password == "") and self.login_url:
```
Instead of empty string should we make username and password optional with default values to None ? This check could be replaced by `if not self.username and not self.password`

If we go this way we could also add an error message to the user if all of login_url, user_name and password are falsy

awennersteen replied:
I think empty string is also `False`y

MatthieuMoreau0 replied:
That's correct I believe! My main point was around defining default values so that we wouldn't need to do username="" when initializing the SDK with the login_url

awennersteen replied:
exactly, and having type consistency there would be a bit easier for me downstreams

MatthieuMoreau0 reviewed (commented):
Just curious : How does the login url work ?

anneclairelh commented:
@MatthieuMoreau0 @awennersteen I addressed your comment and also after discussion with James did a refactor of the client to use authenticators and token providers. This way for jupyter hub the only thing is to provide a custom token provider like so:

```
class JupyterHubTokenProvider(TokenProvider):
    def __init__(self, hub_api_url: str, hub_api_token: str):
        self.hub_api_url = hub_api_url
        self.hub_api_token = hub_api_token

    def _query_token(self) -> str:
        return requests.get(
            self.hub_api_url, headers={"Authorization": f"Bearer {self.hub_api_token}"}
        ).json()["auth_state"]["access_token"]
```

awennersteen commented on sdk/client.py:
```diff
@@ -59,13 +134,33 @@ def _login(self) -> None:
+
+    def __init__(
+        self,
+        group_id: str,
```
why is this one not Optional here?

MatthieuMoreau0 commented on sdk/client.py:
```diff
@@ -49,7 +124,7 @@ def _login(self) -> None:
 
         rsp = requests.post(
-            url,
+            self.login_url,
```
We build a URL above but do not use it ?

anneclairelh replied:
Sorry, remnant from earlier version.

MatthieuMoreau0 commented on sdk/__init__.py:
```diff
@@ -28,13 +28,20 @@ class SDK:
         group_id: str,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        login_url: Optional[str] = None,
```
Should this have a convenient default value for the users ? (Is it even necessary anymore ?)

awennersteen replied:
To me I'd still prefer having the default being "" when possible

anneclairelh replied:
Ha, I've changed everything now... ;) Why so?

awennersteen replied:
It's just an unnecessary type instability (in lack of a better word). For dict, str, list etc it's quite nice (and makes mypy extra happy) to leave them as their empty variants.

On the other hand, it's perhaps more user-instructing to mark them as Optional

awennersteen commented:
Thanks!
Indeed this gives me all that's needed

awennersteen reviewed (approved):
LGTM
